### PR TITLE
Sine movement; Added actions to set the horizontal and vertical progress

### PR DESCRIPTION
**Pull request by tristanbob**
A user requested this feature and I think it is useful also.

## Playable game

https://liluo.io/instant-builds/22eed7d5-6c41-4f15-9705-b1b50aab2947

## Project

[SineMovement-0.0.6.zip](https://github.com/GDevelopApp/GDevelop-extensions/files/9467006/SineMovement-0.0.6.zip)

## Video

https://user-images.githubusercontent.com/8879811/187835369-8d94f1a6-9d31-4747-b559-51d6abe87cb6.mp4

### CODE REVIEW

**Comment by D8H** [comment]
What's the use of CenterPointX and CenterPointY properties?

Some suggestions:
- "Reset sine progress counters" to "Reset at center"?
- A "Sine movement configuration" functions group? to have the 3 progress actions in the default group. 
- Remove the "Set" prefix in actions full name.
- "amplitude instead of "distance".

**Comment by tristanbob** [comment]
> What's the use of CenterPointX and CenterPointY properties?

It saves the starting position (or it can be set manually).  I tried moving this logic to "OnCreated", but that function doesn't know the X() and Y() position of the object because they always return "0".

> Some suggestions:
> 
> * "Reset sine progress counters" to "Reset at center"?

I changed it to "Reset to starting position"

> * A "Sine movement configuration" functions group? to have the 3 progress actions in the default group.

Ok, I think I changed this to what you ask.

> * Remove the "Set" prefix in actions full name.

Done.

> * "amplitude instead of "distance".

Where did you want this changed?  I see the properties has both "distance" and "amplitude" in the description.

**Comment by tristanbob** [comment]
@D8H any other changes for this review?  (Thanks for all your help!)

File: extensions/reviewed/SineMovement.json [hunk 1]
```diff
@@ -1185,6 +1241,7 @@
           "type": "Number",
           "label": "Horizontal distance: amplitude of the movement on X axis (0 to deactivate)",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:1242]:
```suggestion
          "label": "Horizontal amplitude",
```

File: extensions/reviewed/SineMovement.json [hunk 2]
```diff
@@ -1194,6 +1251,7 @@
           "type": "Number",
           "label": "Vertical distance: amplitude of the movement on Y axis (0 to deactivate)",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:1252]:
```suggestion
          "label": "Vertical amplitude",
```

File: extensions/reviewed/SineMovement.json [hunk 3]
```diff
@@ -899,33 +971,29 @@
         },
         {
           "description": "Set vertical distance.",
-          "fullName": "Set vertical distance",
+          "fullName": "Vertical distance",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:974]:
```suggestion
          "fullName": "Vertical amplitude",
```

File: extensions/reviewed/SineMovement.json [hunk 4]
```diff
@@ -899,33 +971,29 @@
...
           "functionType": "Action",
+          "group": "Sine movement configuration",
           "name": "SetVerticalDistance",
           "private": false,
           "sentence": "Set vertical distance of _PARAM0_ to _PARAM2_",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:979]:
```suggestion
          "sentence": "Change the vertical amplitude of _PARAM0_ to _PARAM2_",
```

File: extensions/reviewed/SineMovement.json [hunk 5]
```diff
@@ -899,33 +971,29 @@
         },
         {
           "description": "Set vertical distance.",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:973]:
```suggestion
          "description": "Change vertical amplitude.",
```

File: extensions/reviewed/SineMovement.json [hunk 6]
```diff
@@ -834,33 +910,29 @@
         },
         {
           "description": "Set horizontal distance.",
-          "fullName": "Set horizontal distance",
+          "fullName": "Horizontal distance",
           "functionType": "Action",
+          "group": "Sine movement configuration",
           "name": "SetHorizontalDistance",
           "private": false,
           "sentence": "Set horizontal distance of _PARAM0_ to _PARAM2_",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:918]:
```suggestion
          "sentence": "Change the horizontal amplitude of _PARAM0_ to _PARAM2_",
```

File: extensions/reviewed/SineMovement.json [hunk 7]
```diff
@@ -834,33 +910,29 @@
         },
         {
           "description": "Set horizontal distance.",
-          "fullName": "Set horizontal distance",
+          "fullName": "Horizontal distance",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:913]:
```suggestion
          "fullName": "Horizontal amplitude",
```

File: extensions/reviewed/SineMovement.json [hunk 8]
```diff
@@ -834,33 +910,29 @@
         },
         {
           "description": "Set horizontal distance.",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:912]:
```suggestion
          "description": "Change the horizontal amplitude.",
```

File: extensions/reviewed/SineMovement.json [hunk 9]
```diff
@@ -550,28 +524,24 @@
           "description": "Vertical distance.",
           "fullName": "Vertical distance",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:525]:
```suggestion
          "fullName": "Vertical amplitude",
```

File: extensions/reviewed/SineMovement.json [hunk 10]
```diff
@@ -550,28 +524,24 @@
           "description": "Vertical distance.",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:524]:
```suggestion
          "description": "Return the vertical amplitude.",
```

File: extensions/reviewed/SineMovement.json [hunk 11]
```diff
@@ -498,28 +476,24 @@
           "description": "Horizontal distance.",
           "fullName": "Horizontal distance",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:477]:
```suggestion
          "fullName": "Horizontal amplitude",
```

File: extensions/reviewed/SineMovement.json [hunk 12]
```diff
@@ -498,28 +476,24 @@
           "description": "Horizontal distance.",
```
Comment (D8H) [inline extensions/reviewed/SineMovement.json:476]:
```suggestion
          "description": "Return the horizontal amplitude.",
```

**Comment by D8H** [comment]
> > What's the use of CenterPointX and CenterPointY properties?
> 
> It saves the starting position (or it can be set manually). I tried moving this logic to "OnCreated", but that function doesn't know the X() and Y() position of the object because they always return "0".

The behavior should probably watch position changes done by the outside and update the center accordingly.
If the property is only hidden (but keeping the action and expression), I think it's still fine. I can't picture a user setting the property when they can just place the object.
SineProgressX and SineProgressX could be made visible as initial values. Should they be between 0 and 1 or be called angle?

**Comment by tristanbob** [comment]
>

**Comment by tristanbob** [comment]
> The behavior should probably watch position changes done by the outside and update the center accordingly. 

This would be a nice feature, but it is not simple and I don't want to add that scope to this PR. 

> If the property is only hidden (but keeping the action and expression), I think it's still fine. I can't picture a user setting the property when they can just place the object. 

A user requested it, and I think my example that tracked mouse position could make some fun visual effects.

> SineProgressX and SineProgressY could be made visible as initial values. Should they be between 0 and 1 or be called angle?

If I was rewriting this from scratch, I would hide degrees from the user completely.  They just need to know:
- How far will it move in each direction (pixels)
- How long it takes to complete a full cycle (seconds)

Since we are not creating this from scratch, I'm not sure how to improve it. I wasn't planning on a full refactor when adding these additional functions that a user requested.

**Comment by D8H** [comment]
> > If the property is only hidden (but keeping the action and expression), I think it's still fine. I can't picture a user setting the property when they can just place the object.
> 
> A user requested it, and I think my example that tracked mouse position could make some fun visual effects.

It don't understand. It doesn't need the property CenterPointX and CenterPointY to be visible in the configuration, does it?

> > SineProgressX and SineProgressY could be made visible as initial values. Should they be between 0 and 1 or be called angle?
> 
> If I was rewriting this from scratch, I would hide degrees from the user completely. They just need to know:
> 
>     * How far will it move in each direction (pixels)
> 
>     * How long it takes to complete a full cycle (seconds)
> 
> 
> Since we are not creating this from scratch, I'm not sure how to improve it. I wasn't planning on a full refactor when adding these additional functions that a user requested.

The "progress" wasn't visible to extension users before, right?
If this is the case, now is the right time to change it because after this update it won't be possible anymore without a breaking change.

**Comment by tristanbob** [comment]
> It don't understand. It doesn't need the property CenterPointX and CenterPointY to be visible in the configuration, does it?

This is the only way the user can move an object with this behavior.  If they try moving the object, it gets instantly teleported to it's path based on the center value.  The only way to remove this is to only apply deltas (changes in position) instead of absolute position.  I might be able to figure out a way to do this, but I don't have any ideas at the moment.  (This is how the object shake extension tracks movement)

> The "progress" wasn't visible to extension users before, right? If this is the case, now is the right time to make the change because after this update it won't be possible anymore without a breaking change.

Yes, I agree, however, the speed and distance (amplitude) parameters did exist, so we can't change those.

**Comment by D8H** [comment]
> > It don't understand. It doesn't need the property CenterPointX and CenterPointY to be visible in the configuration, does it?
> 
> This is the only way the user can move an object with this behavior. If they try moving the object using normal methods, the object gets instantly teleported along it's sine-path based on the center value. The only way to remove this is to only apply deltas (changes in position) instead of absolute position. I might be able to figure out a way to do this, but I don't have any ideas at the moment. (This is how the object shake extension tracks movement)

When this property is left to 0, it's set to the object position at the beginning. I'm just saying that it should be hidden to always do this. It doesn't change anything in the events.

> > The "progress" wasn't visible to extension users before, right? If this is the case, now is the right time to make the change because after this update it won't be possible anymore without a breaking change.
> 
> Yes, I agree, however, the speed and distance (amplitude) parameters did exist, so we can't change those.

Ok, the speed is expressed in angle per second. It should have been a duration for one cycle but well.
But, it doesn't forbid the "progress" to be between 0 and 1.

**Comment by D8H** [comment]
I think we should make a breaking change to make it an ellipse (like it was initially) and allow to move the object this way:
- both amplitude != 0 => ellipse
- both amplitude != 0 + a force => loops
- one amplitude != 0 => back and forth
- one amplitude != 0 + force => waves

**Comment by tristanbob** [comment]
> I think we should make a breaking change to make it an ellipse (with only one speed property) and allow to move the object this way:

I do not want to remove the ability to independently set the X and Y speed.  This feature provides the most flexibility for users to create a variety of movement types (back and forth, circle, figure 8, chaotic, etc)

> There could even be a property to rotate the object to face forward.

I thought about this, but I'd rather have people add the FaceForward extension to the object.

**Comment by tristanbob** [comment]
> When this property is left to 0, it's set to the object position at the beginning. I'm just saying that it should be hidden to always do this. It doesn't change anything in the events.

Yes, I will hide this property but allow users to set it using events.

> Ok, the speed is expressed in angle per second. It should have been a duration for one cycle but well. But, it doesn't forbid the "progress" to be between 0 and 1.

I'll investigate and try to implement this recommendation.

**Comment by D8H** [comment]
> > I think we should make a breaking change to make it an ellipse (with only one speed property) and allow to move the object this way:
> 
> I do not want to remove the ability to independently set the X and Y speed. This feature provides the most flexibility for users to create a variety of movement types (back and forth, circle, figure 8, chaotic, etc)

With `ProgressX` and `ProgressY`, how can users make sense of the actual progress?
A full cycle (both progress to 0) may even never happen.
A chaotic movement should be the role of another extension, a noise-based ShakeObject perhaps. If people want a strange rose movement, they can still add this extension twice or do this very specific thing themselves.
I bet that this extension is used with `speedX = speedY`, 90% of the time. Why make things harder for the most?

With only one speed, the "progress" can just be a angle which is easy to picture.
People could use an action "Turn the other way" that would toggle the center on the left or right side of the object. Repeated every 180°, it allows to do nice curves like this:
Arc

**Comment by D8H** [comment]
A made a version of the extension.

New features:
- The object starts from where it was on the scene.
- The initial angle and direction can be setup with properties
- The object can turn left and right
- The object can be move by another behavior (force, screen warping...)

I made a mini game in only 2 events to illustrate:

image

The build:
https://liluo.io/games/be98d949-c9a0-4972-86df-1cfb5a10fea6

Its project:
https://www.dropbox.com/s/g55dyp1vj5gtlm7/EllipseMovementExample.zip?dl=0

**Comment by tristanbob** [comment]
This extension has been deprecated and replaced with a much more versatile version: `Ellipse movement` https://github.com/GDevelopApp/GDevelop-extensions/pull/591

**Pull request closed without merging**